### PR TITLE
Update correct next approver with category/tag rules

### PR DESCRIPTION
**Pull request by nkdengineer**
### Explanation of Change

- [x] Update `getApprovalChain` to include category/tag rule approvers
- [x] `getNextApproverAccountID` already works as expected

### Fixed Issues

$ https://github.com/Expensify/App/issues/52458
PROPOSAL:

### Tests

Precondition:
- The workspace's rules are enabled, and the approval mode is `ADVANCE`
- Has 4 approvers and 1 submitter
- Setting two category approver rules (CAT1 and CAT2) for two approvers (approver 1 and approver 2) and two tag approver rules (TAG1 and TAG2) for two approvers (approver 3 and approver 4)

1. [Submitter]: Submit 4 expenses that include categories `CAT1` and `CAT2`, tags `TAG1` and `TAG2`
2. [Submitter]: Submit the expense report and verify that the next approver is the approver 1
3. [Approver1]: Approve the expense and verify that the next approver is the approver 2
4. [Approver2]: Approve the expense and verify that the next approver is the approver 3
5. [Approver3]: Approve the expense and verify that the next approver is the approver 4
6. [Approver4]: Approve the expense and verify that the next step is waiting admin for payment

- With the same setting above, disable the approval
1. [Submitter]: Create an expense and submit it
2. Verify that no approval is needed and the expense is submitted successfully
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above
### QA Steps

// TODO: These must be filled out, or the issue title must include "[No QA]."
Precondition:
- The workspace's rules are enabled, and the approval mode is `ADVANCE`
- Has 4 approvers and 1 submitter
- Setting two category approver rules (CAT1 and CAT2) for two approvers (approver 1 and approver 2) and two tag approver rules (TAG1 and TAG2) for two approvers (approver 3 and approver 4)

1. [Submitter]: Submit 4 expenses that include categories `CAT1` and `CAT2`, tags `TAG1` and `TAG2`
3. [Submitter]: Submit the expense report and verify that the next approver is the approver 1
4. [Approver1]: Approve the expense and verify that the next approver is the approver 2
5. [Approver2]: Approve the expense and verify that the next approver is the approver 3
6. [Approver3]: Approve the expense and verify that the next approver is the approver 4
7. [Approver4]: Approve the expense and verify that the next step is waiting admin for payment
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/34fb2353-655f-40ce-88db-d2b3d4a550c6

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/85d22fd8-4b5f-4750-8d75-7a9a4d2104d5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/de64e02e-8921-4d0a-9b1a-3ae96e063746

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/24f24600-7e31-45fa-b7f9-28ffaebe3fc2

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/25921685-eebc-4104-b3ab-ce9940fbb6b7

https://github.com/user-attachments/assets/a5408778-9fe9-4717-af2c-a1a8c2586197

https://github.com/user-attachments/assets/16c60d7e-1388-4b03-a514-66e09f65ad07

https://github.com/user-attachments/assets/21eace33-6a23-416f-9a66-c149ba0cfe28

https://github.com/user-attachments/assets/a0c3e98f-e97f-4464-bc6d-79b012edbf94

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/f43bc32e-1db7-4c3a-a9ef-4a6e9ce08e9e

</details>

### CODE REVIEW

**Comment by Beamanator** [comment]
I spent a lot of time today fixing the backend, PR is up - but I didn't get to test your front end changes, sorry - if you can't find a bug with the correct approver when there's multiple category approvers, that's great! I will try to reproduce the issue i found on Monday 🙏

**Comment by nkdengineer** [comment]
Got it, no problem.

File: src/libs/TransactionUtils/index.ts
Comment (nkdengineer) [reply src/libs/TransactionUtils/index.ts:1238]: > // We need sort all transactions by sorting the parent report actions because `created` of the transaction only has format `YYYY-MM-DD` which can cause the wrong sorting

@Beamanator Currently, we're going to the first expense. I think I need to update the comment a bit to 

```
// We need to sort all transactions by sorting the parent report actions because `created` of the transaction doesn't mean the created time of the transaction.
```

**Comment by nkdengineer** [comment]
@ntdiary Is this reproducible on the latest main? If yes, I think we should fix it as a separate issue.

File: src/libs/TransactionUtils/index.ts
Comment (Beamanator) [reply src/libs/TransactionUtils/index.ts:1238]: Ok well from my testing in OldDot, it honestly doesn't seem consistent, which order the category approvers show up when there's multiple 😅  So I'm discussing internally to figure out what we'll want to do in NewDot, as an expected order

Comment (Beamanator) [reply src/libs/TransactionUtils/index.ts:1238]: Ok Updated sort order is here: https://github.com/Expensify/App/issues/52458#issuecomment-2504379629

Comment (nkdengineer) [reply src/libs/TransactionUtils/index.ts:1238]: Still investigate the sort order.

**Comment by Beamanator** [comment]
I definitely think it's fair to bump the bounty on this one, it's a pretty complicated issue that we're trying to get perfect 👍 

@ntdiary are you waiting on me or @nkdengineer to look at your latest message?

BTW I'm making sure a transaction `inserted` date gets back to Onyx, that should be ready in staging tomorrow at some point 🙏

**Comment by Beamanator** [comment]
Cool, yeah @nkdengineer can now implement the correct sorting (with transaction `inserted` date - b/c that was just deployed yesterday)

> and to see if https://github.com/Expensify/App/pull/52537#issuecomment-2504308501 should be fixed. I personally think it should be fixed, because the expected behavior was mentioned in the OP, and in our case, the frontend didn't correctly add the owner to the full approval chain(i.e., rule approval chain + submitsTo/owner/forwards). 😄

I think this makes sense too, to fix here 👍 i believe it should be a relatively quick fix 😅

**Comment by Beamanator** [comment]
Ooh definitely agreed about tests 👍  though i wouldn't mind if that's in a follow-up just so we can get this out the door quicker 🤷

**Comment by nkdengineer** [comment]
> Maybe this should be fixed as a frontend bug: When no.4 approves, the frontend optimistic message shows no further action required!, but the correct backend data is waiting for no.25 to approve. (no.25 is policy owner)

@Beamanator For this bug, I see it's expected [here](https://github.com/Expensify/App/pull/42018) that will show the `No further action required!` if the approver isn't the payer.

File: src/libs/TransactionUtils/index.ts
Comment (nkdengineer) [reply src/libs/TransactionUtils/index.ts]: @ntdiary Updated the sort order.

**Comment by Beamanator** [comment]
Ok IDK bout y'all, but I think y'all need to be suuuuuper clear on the setup you're using to confirm / deny if the policy owner is expected to be the "final approver" in your cases.

@nkdengineer the PR you linked doesn't have anything to do with category/tag approvers, it looks like a super simple setup - does that apply here?

Do you both understand each other's setups or should you both lay out the setup (policy type, `autoReportingFrequency`, approval mode (basic vs advanced), members list / who approves to who, category & tag approvers, order of expense creation w/ their category/tags)

**Comment by Beamanator** [comment]
Thanks @ntdiary that's very helpful! Can you also please show what the approval workflow looks like? I see you mention "the final approver is policy owner (no.25)" but I would love to see how that's set up in your `employeeList` just to confirm 🙏  everything else makes complete sense!

**Comment by Beamanator** [comment]
Thanks @ntdiary ! 👍 

Ok ya so it IS expected that, in your case, the policy owner is the final approver - this is because the submitter's `submitsTo` is the policy owner, not because we want to always end with policy owner 👍

**Comment by nkdengineer** [comment]
@Beamanator So it's not a bug, right?

**Comment by Beamanator** [comment]
Right, in @ntdiary 's case at least, it's not a backend bug, it is expected for the `submitsTo` to approve (in advanced approval case) AFTER all of the rule approvers

**Comment by Beamanator** [comment]
Ya thanks for getting us full circle 😂 I honestly do think this is
something we need to fix in this issue - agreed we don’t want this to be
counted as a regression later

**Comment by Beamanator** [comment]
@ntdiary yikes, yeah in OldDot I'm struggling to even get multi-level tags set up in a policy 🙈  

I guess NewDot doesn't support multiple tags even at the moment, so probably won't support multiple tag approvers for a bit, so I would say we can skip that part of tests / sorting & add it in later whenever that's added to the app

**Comment by Beamanator** [comment]
So what's the status of this PR? @nkdengineer are you waiting on anyone for anything?

File: src/libs/TransactionUtils/index.ts
Comment (nkdengineer) [reply src/libs/TransactionUtils/index.ts:1255]: I think we can make it optional and update it in `buildOptimisticTransaction` because the draft transaction don't need to add this field.

Comment (nkdengineer) [reply src/libs/TransactionUtils/index.ts:1256]: @ntdiary I updated.

**Comment by Beamanator** [comment]
Bump @ntdiary did you have any chance to test today?

**Comment by Beamanator** [comment]
Oof!!! Feel free to rest & pick this up tomorrow!

**Review by ntdiary — CHANGES_REQUESTED** [review]
> Oof!!! Feel free to rest & pick this up tomorrow!

@Beamanator, thank you very much! I have took a quick look, and the code seems to be working fine. However, I’ve also suggested several changes, hope to make the code clearer, will continue testing it tomorrow. :D

cc @nkdengineer

**Comment by nkdengineer** [comment]
@ntdiary I updated.

**Review by ntdiary — CHANGES_REQUESTED** [review]
@nkdengineer, Thank you very much! We still have some lint errors to fix and unused code to remove, can you please address them? The changes are minor, so I’ve already recorded the test video and am filling the check list, will approve soon. :D

**Review by ntdiary — APPROVED** [review]
LGTM!
The errors reported by ESLint are from code that we haven’t modified. :)

**Comment by nkdengineer** [comment]
@Beamanator Let's wait a bit. I just found a case that we cannot submit the report when the approval flow is disabled.

File: src/libs/ReportUtils.ts
Comment (nkdengineer) [inline src/libs/ReportUtils.ts:8479]: @ntdiary Clarify a bit, we need to return early if the approval mode is `OPTIONAL` then when the user submits the report, backend will not return error about the `managerAccountID` is incorrect.

Comment (nkdengineer) [reply src/libs/ReportUtils.ts:8479]: You can test with the case the step-up is the same with the original step except the approval mode is `OPTIONAL`. Then `no.24` creates an expense with tag/category match with rule and submit the report.

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8479]: Ooh ok I will test that out in OldDot now to see what happens!

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8479]: OK I just tested in OldDot and found that Category Approvers **are NOT** supported on Control, Submit & Close policies 🙏  - I honestly had never tested that, but glad it's tested now & it sounds like that's working in this PR 🙏

Comment (Beamanator) [reply src/libs/ReportUtils.ts:8479]: > Interesting, @nkdengineer, so you tried this case:
> 
> 1. enable `Advanced Approval` mode
> 2. set category&tag rule approvers
> 3. then switch to `Submit and Close` mode directly.

FYI I don't think it's important to run the test in this order - a.k.a. it's not important to enable advanced approval first, then switch to submit & close... Right?!?

**Comment by Beamanator** [comment]
@nkdengineer conflicts 🙏  I'm doing a final run through today, but seems we might be able to get this merged today!!

**Comment by nkdengineer** [comment]
Update now.

File: src/libs/PolicyUtils.ts
Comment (Beamanator) [inline src/libs/PolicyUtils.ts:570]: Here we're getting the NEXT manager account ID, right? Can we call this function something like `getNextManagerAccountID` & rename `getManagerAccountEmail` to something like `getNextManagerAccountEmail`?

The way I currently read these functions is that we should be getting the report's current manager account ID / email - which I don't believe is true

File: src/libs/ReportUtils.ts
Comment (Beamanator) [inline src/libs/ReportUtils.ts:8488]:
```suggestion
        // Don't push submitter to approve as a rule approver
```

**Comment by nkdengineer** [comment]
@Beamanator Resolved conflict.

**Review by Beamanator — APPROVED** [review]
Ok looooooking good, just a few comments:
1. So far the `QA Steps` only include tests for the `Advanced` approval mode. Please update to include tests for `Submit & Approve` and `Submit & Close` 🙏 
    - I think it would be also smart to make a quick note that QA should test approval modes on Collect plans as well to make sure they're still working perfect
2. There's actually another part which I think we are missing - which is basically: when building the approval chain, we should check if everyone in the chain has already approved or not - because it's possible for the order to get mixed up if someone approves, then the approval chain changes. This, however, I think I'll try to handle in a follow-up cuz it's a pretty annoying case & we should get this one out the door 😅

**Comment by Beamanator** [comment]
Failing test was esLint errors unrelated to these changes, so not an `Emergency`

**Comment by Beamanator** [comment]
Failing test was esLint errors unrelated to these changes, so not an `Emergency`

**Pull request merged**
